### ISSUE TITLE
Startup-Error: IllegalStateException: GlobalOpenTelemetry.set has already been called. 

### ISSUE THREAD

**Issue by matthiaswronka** [quarkusio/quarkus#22805]
### Describe the bug

Sometimes (5 % of Pipelines) we build Docker-Images that fail to start due to an initialization-error in OpenTelemetry:

```
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 15 more
Caused by: java.lang.IllegalStateException: GlobalOpenTelemetry.set has already been called. GlobalOpenTelemetry.set must be called only once before any calls to GlobalOpenTelemetry.get. If you are using the OpenTelemetrySdk, use OpenTelemetrySdkBuilder.buildAndRegisterGlobal instead. Previous invocation set to cause of this exception.
	at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:84)
	at io.opentelemetry.sdk.OpenTelemetrySdkBuilder.buildAndRegisterGlobal(OpenTelemetrySdkBuilder.java:63)
	at io.quarkus.opentelemetry.runtime.OpenTelemetryRecorder.createOpenTelemetry(OpenTelemetryRecorder.java:33)
	at io.quarkus.deployment.steps.OpenTelemetryProcessor$createOpenTelemetry1944204174.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.OpenTelemetryProcessor$createOpenTelemetry1944204174.deploy(Unknown Source)
	... 16 more
Caused by: java.lang.Throwable
	at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:92)
	at io.opentelemetry.api.GlobalOpenTelemetry.get(GlobalOpenTelemetry.java:62)
	at io.opentelemetry.api.GlobalOpenTelemetry.getTracer(GlobalOpenTelemetry.java:111)
	at io.quarkus.opentelemetry.runtime.tracing.TracerProducer.getTracer(TracerProducer.java:2
```

We build a jar-file using `gradle jar quarkusBuild` and then add this to a Docker-image. If the error occours, it occours every time the image is used (on developer machine, in the cloud, ... doesn´t matter). If we re-run the pipeline with nearly no code-change (we added a commit with an extra `space` in a readme) chances are that the error disappears.

What might be application-specific is a Tracing-Interceptor we use. That gets the `Tracer` injected:

```kotlin
@Provider
@Priority(Interceptor.Priority.APPLICATION)
class TracingInterceptor : ContainerResponseFilter {

    override fun filter(ctx: ContainerRequestContext, responseContext: ContainerResponseContext) {
        val spanName = ctx.method + " " + ctx.uriInfo.path
        val span = Span.current() ?: return
        span.updateName(spanName)
        val traceId = span.spanContext?.traceId
        MDC.put("traceId", traceId)
        responseContext.headers.putSingle("TRACE-ID", traceId ?: "no tracing")
    }
}

@Interceptor
@Tracing
@Suppress("TooGenericExceptionCaught")
@Priority(Interceptor.Priority.APPLICATION)
class TracingMethodInterceptor {

    @Inject
    lateinit var tracer: Tracer

    @Inject
    lateinit var request: RequestAttributes

    @ConfigProperty(name = "tracing.logParameters")
    var logParameters: Boolean = true

    @AroundInvoke
    @Throws(Exception::class)
    fun trace(ctx: InvocationContext): Any? {
        val parentSpan = Span.current()
        request.attributes["traceId"] = parentSpan.spanContext.traceId

        val methodDisplay = "${ctx.method.name} ${ctx.method.declaringClass.simpleName}"
        val fullMethodName = ctx.method.toGenericString()
        val span = tracer.spanBuilder(methodDisplay)
            .setParent(Context.current().with(parentSpan))
            .startSpan()

        span.makeCurrent().use {
            span.setAttribute("method", fullMethodName)

            if (logParameters) {
                val parameterDisplay = ctx.parameters.filterNotNull().joinToString { "$it" }
                span.setAttribute("parameters", parameterDisplay)
            }

            try {
                return ctx.proceed()
            } catch (e: Exception) {
                e.logToTracer()
                throw e
            } finally {
                span.end()
            }
        }
    }
}
```

By now we consider this to be a build-problem, since the error is tight to a specific artifact and not to the code or the runtime-environment.

What might be of interest:
- we use kotlin
- we use gradle
- we do not use graalvm

### Expected behavior

Every build of the same codebase should deliver reproduceable results.

### Actual behavior

On a random basis the error described above occours. We do not have any hint why this happens.

### How to Reproduce?

Unfortunately we do not understand the error enough to provide a reproduction-instruction. Maybe you have an idea what might be the problem from the information we provided or you have a hint where we could dig deeper.

### Output of `uname -a` or `ver`

Linux runner-t9jgdck6-project-9010693-concurrent-0 5.11.0-1026-gcp #29~20.04.1-Ubuntu SMP Fri Jan 7 12:24:31 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.13" 2021-10-19   
OpenJDK Runtime Environment Temurin-11.0.13+8 (build 11.0.13+8)  
OpenJDK 64-Bit Server VM Temurin-11.0.13+8 (build 11.0.13+8, mixed mode)

### Quarkus version or git rev

io.quarkus:quarkus-bom:2.6.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
------------------------------------------------------------
Gradle 7.3.3
------------------------------------------------------------
Build time:   2021-12-22 12:37:54 UTC
Revision:     6f556c80f945dc54b50e0be633da6c62dbe8dc71
Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.13 (Eclipse Adoptium 11.0.13+8)
OS:           Linux 5.11.0-1026-gcp amd64
```

### Additional information

The Environment-Infos describe the Build-Environment.

**Comment by radcortez** [comment]
I know that this is not related, but out of curiosity any particular reason for adding the custom tracer interceptors? In Quarkus, REST Endpoints are automatically traced, and for CDI Beans you can use the OTEL `@WithSpan` annotation to trace method calls. If something is missing here, we can certainly add it.

Regarding the issue, can you confirm the Quarkus version is 2.6.1.Final? Because the stacktrace doesn't seem to match the current source, in:

```
at io.opentelemetry.sdk.OpenTelemetrySdkBuilder.buildAndRegisterGlobal(OpenTelemetrySdkBuilder.java:63)
at io.quarkus.opentelemetry.runtime.OpenTelemetryRecorder.createOpenTelemetry(OpenTelemetryRecorder.java:33)
```

This was changed in https://github.com/quarkusio/quarkus/commit/57f6a9d9cb288cce2b558d793948b8d0e56ea806#diff-eeba33b113647277cc8169f3f0550e0bc38d070c7c367586be30d308bb7e7d67L33, which was added in 2.6.0.Final. 

The call to `buildAndRegisterGlobal` should be line 35: https://github.com/quarkusio/quarkus/blob/2.6.1.Final/extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryRecorder.java#L35

**Comment by matthiaswronka** [comment]
Let me answer the second part first: configuration in `build.gradle` seems fine and dependencies are resolved as follows.

```
$>gradle dependencies | grep quarkus
+--- io.quarkus:quarkus-bom:2.6.1.Final
|    +--- io.quarkus:quarkus-agroal:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-config-yaml:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-hibernate-orm:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-hibernate-validator:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-jackson:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-jdbc-postgresql:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-keycloak-admin-client:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-kotlin:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-narayana-jta:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-oidc:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-opentelemetry-exporter-otlp:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-quartz:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-rest-client:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-resteasy-jackson:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-resteasy-multipart:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-resteasy:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-scheduler:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-security:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-undertow:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-core:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-arc:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-datasource:2.6.1.Final (c)
|    +--- io.quarkus:quarkus-credentials:2.6.1.Final (c)
...
```

I will double check, if these files make it into the failing jar.

**Comment by matthiaswronka** [comment]
I´m sorry. I accidently picked an older stacktrace. The problem stays the same but now with a current stacktrace (see below) the line-numbers match the expectations (and yes, we double checked and confirm to use 2.6.1.Final):

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:53)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:28)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:66)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 15 more
Caused by: java.lang.IllegalStateException: GlobalOpenTelemetry.set has already been called. GlobalOpenTelemetry.set must be called only once before any calls to GlobalOpenTelemetry.get. If you are using the OpenTelemetrySdk, use OpenTelemetrySdkBuilder.buildAndRegisterGlobal instead. Previous invocation set to cause of this exception.
	at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:84)
	at io.opentelemetry.sdk.OpenTelemetrySdkBuilder.buildAndRegisterGlobal(OpenTelemetrySdkBuilder.java:63)
	at io.quarkus.opentelemetry.runtime.OpenTelemetryRecorder.createOpenTelemetry(OpenTelemetryRecorder.java:35)
	at io.quarkus.deployment.steps.OpenTelemetryProcessor$createOpenTelemetry1944204174.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.OpenTelemetryProcessor$createOpenTelemetry1944204174.deploy(Unknown Source)
	... 16 more
Caused by: java.lang.Throwable
	at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:92)
	at io.opentelemetry.api.GlobalOpenTelemetry.get(GlobalOpenTelemetry.java:62)
	at io.opentelemetry.api.GlobalOpenTelemetry.getTracer(GlobalOpenTelemetry.java:111)
	at io.quarkus.opentelemetry.runtime.tracing.TracerProducer.getTracer(TracerProducer.java:30)
	at io.quarkus.opentelemetry.runtime.tracing.TracerProducer_ProducerMethod_getTracer_d966e02c84dcc89f26dde2c4294d174ee3b55ece_Bean.create(Unknown Source)
	at io.quarkus.opentelemetry.runtime.tracing.TracerProducer_ProducerMethod_getTracer_d966e02c84dcc89f26dde2c4294d174ee3b55ece_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.opentelemetry.runtime.tracing.TracerProducer_ProducerMethod_getTracer_d966e02c84dcc89f26dde2c4294d174ee3b55ece_Bean.get(Unknown Source)
	at io.quarkus.opentelemetry.runtime.tracing.TracerProducer_ProducerMethod_getTracer_d966e02c84dcc89f26dde2c4294d174ee3b55ece_Bean.get(Unknown Source)
	at com.helsinki.interception.tracing.TracingMethodInterceptor_Bean.create(Unknown Source)
	at com.helsinki.interception.tracing.TracingMethodInterceptor_Bean.get(Unknown Source)
	at com.helsinki.interception.tracing.TracingMethodInterceptor_Bean.get(Unknown Source)
	at com.helsinki.externalApi.logging.FileExternalApiLogRepository_Subclass.<init>(Unknown Source)
	at com.helsinki.externalApi.logging.FileExternalApiLogRepository_Bean.create(Unknown Source)
	at com.helsinki.externalApi.logging.FileExternalApiLogRepository_Bean.get(Unknown Source)
	at com.helsinki.externalApi.logging.FileExternalApiLogRepository_Bean.get(Unknown Source)
	at com.helsinki.externalApi.logging.ExternalApiLoggingFilter_Bean.create(Unknown Source)
	at com.helsinki.externalApi.logging.ExternalApiLoggingFilter_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at com.helsinki.externalApi.logging.ExternalApiLoggingFilter_Bean.get(Unknown Source)
	at com.helsinki.externalApi.logging.ExternalApiLoggingFilter_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:440)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:453)
	at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:277)
	at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:274)
	at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:39)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.injectedInstance(ResteasyProviderFactoryImpl.java:1399)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$AbstractInterceptorFactory.createInterceptor(JaxrsInterceptorRegistryImpl.java:150)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.initialize(JaxrsInterceptorRegistryImpl.java:168)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.checkInitialize(JaxrsInterceptorRegistryImpl.java:183)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.getInterceptor(JaxrsInterceptorRegistryImpl.java:193)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$AbstractInterceptorFactory.postMatch(JaxrsInterceptorRegistryImpl.java:124)
	at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl.postMatch(JaxrsInterceptorRegistryImpl.java:288)
	at org.jboss.resteasy.core.ResourceMethodInvoker.<init>(ResourceMethodInvoker.java:143)
	at org.jboss.resteasy.core.ResourceMethodRegistry.processMethod(ResourceMethodRegistry.java:381)
	at org.jboss.resteasy.core.ResourceMethodRegistry.register(ResourceMethodRegistry.java:308)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:259)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:227)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:208)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:192)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:175)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addPerRequestResource(ResourceMethodRegistry.java:87)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.registerResources(ResteasyDeploymentImpl.java:518)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.registration(ResteasyDeploymentImpl.java:475)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.startInternal(ResteasyDeploymentImpl.java:164)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.start(ResteasyDeploymentImpl.java:121)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.init(ServletContainerDispatcher.java:144)
	at org.jboss.resteasy.plugins.server.servlet.FilterDispatcher.init(FilterDispatcher.java:47)
	at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:112)
	at io.undertow.servlet.core.ManagedFilter.createFilter(ManagedFilter.java:80)
	at io.undertow.servlet.core.DeploymentManagerImpl$2.call(DeploymentManagerImpl.java:591)
	at io.undertow.servlet.core.DeploymentManagerImpl$2.call(DeploymentManagerImpl.java:556)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:42)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$9$1.call(UndertowDeploymentRecorder.java:566)
	at io.undertow.servlet.core.DeploymentManagerImpl.start(DeploymentManagerImpl.java:598)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder.bootServletContainer(UndertowDeploymentRecorder.java:517)
	at io.quarkus.deployment.steps.UndertowBuildStep$build955877346.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.UndertowBuildStep$build955877346.deploy(Unknown Source)
```

**Comment by matthiaswronka** [comment]
> I know that this is not related, but out of curiosity any particular reason for adding the custom tracer interceptors? In Quarkus, REST Endpoints are automatically traced, and for CDI Beans you can use the OTEL @WithSpan annotation to trace method calls. If something is missing here, we can certainly add it.

Thank you. We will check out `@WithSpan`. This might make things easier. If we should miss something we would open a feature request.

**Comment by radcortez** [comment]
> I´m sorry. I accidently picked an older stacktrace. The problem stays the same but now with a current stacktrace (see below) the line-numbers match the expectations (and yes, we double checked and confirm to use 2.6.1.Final):

No worries, was just double-checking :)

It seems that the `TracerProducer` is trying to set the `OpenTelemetry` instance again when it was already set. Not sure why. Let me try to dig into it.

**Comment by geoand** [comment]
Any chance we can get a small project that exhibits the problematic behavior?

It would save us time trying to reproduce the problem and allow us to go straight to fixing it.

Thanks

**Comment by radcortez** [comment]
Closing this issue due to lack of activity and reproducer.

**Comment by AndreasPetersen** [comment]
Using Quarkus 2.13.2, setting `@WithSpan` on a static method results in this error.

**Comment by radcortez** [comment]
The `@WithSpan` annotation is not supported in static methods for CDI Beans.

**Comment by sujalmandal** [comment]
I was having the same issue with @WithSpan on methods in repository classes which implements PanacheRepository

**Comment by sujalmandal** [comment]
I was having the same issue with @WithSpan on methods in repository classes which implements PanacheRepository

**Comment by radcortez** [comment]
@sujalmandal can you please, open a new issue with a reproducer? Thank you!
